core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+    """Predict classes for each program in dataset using KNN.
+
+    Args:training_data
+        training_data: dict. A dictionary containing training data.
+            Structure of training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program.
+                        'tokens': list(str). Token list of program.
+                        'k_gram_hashes': list(int). K-gram hash values of
+                            program.
+                        'fingerprint': list(int). Extracted fingerprint of
+                            program.
+                        'top_similar': list(int). Most similar programs to this
+                            program.
+                    } for each program in training_data.
+                }
+        top_neighbours: int. No. of nearest neighbours to consider for KNN.
+
+    Returns:
+        float. Average no. of times a class has to appear in nearest neighbour
+            for it to be correct prediction.
+        list(int). IDs of training_data points which are misclassified by KNN.
+    """
+    training_data = add_top_similars(training_data, top_neighbours)
+
+    # No. of times a class has to appear in nearest neighbours of a program
+    # so that prediction is correct.
+    occurrence = 0
+
+    # Keep record of misclassified programs' ID.
+    misclassified_points = []
+
+    for pid in training_data:
+        similars = training_data[pid]['top_similar']
+        if pid in similars:
+            similars.remove(pid)
+        nearest_classes = [training_data[i]['class'] for (i, _) in similars]
+        cnt = collections.Counter(nearest_classes)
+        common = cnt.most_common(1)
+        training_data[pid]['prediction'] = common[0][0]
+
+        if training_data[pid]['prediction'] == training_data[pid]['class']:
+            # If prediction is correct then add no. of times the correct
+            # class has to appear in nearest neighbours to occurrence.
+            occurrence += common[0][1]
+        else:
+            # Else mark the ID fo program so that it will be
+            # trained using SVM later.
+            misclassified_points.append(pid)
+
+    # Calculate average of occurrence. This is used during prediction so
+    # that if winner of nearest neighbours appears less than the occurrence
+    # we consider that KNN has failed in prediction.
+    occurrence /= float(len(training_data))
+
+    return occurrence, misclassified_points
+
+
+class CodeClassifier(base.BaseClassifier):
+    """A class for code classifier that uses supervised learning to match
+    Python programs to an answer group. The classifier trains on programs
+    that exploration editors have assigned to an answer group.
+    """
+    def __init__(self):
+        super(CodeClassifier, self).__init__()
+        self.training_data = None
```

Comment:
Add single line comments to explain structure, format, and meaning of each of these (feel free to be verbose). I have no idea what self.clf means, for example.